src/views/connectionUI.ts

```diff
@@ -456,25 +456,72 @@ export class ConnectionUI {
     /**
      * Validate a connection profile by connecting to it, and save it if we are successful.
      */
-    public validateAndSaveProfile(profile: Interfaces.IConnectionProfile): PromiseLike<Interfaces.IConnectionProfile> {
+    public validateAndSaveProfile(profile: IConnectionProfile): Promise<IConnectionProfile> {
         const self = this;
         let uri = self.vscodeWrapper.activeTextEditorUri;
         if (!uri || !self.vscodeWrapper.isEditingSqlFile) {
             uri = ObjectExplorerUtils.getNodeUriFromProfile(profile);
         }
-        return self.connectionManager.connect(uri, profile).then(result => {
+        return self.connectionManager.connect(uri, profile).then(async (result) => {
             if (result) {
                 // Success! save it
                 return self.saveProfile(profile);
             } else {
-                // Error! let the user try again, prefilling values that they already entered
-                return self.promptForRetryCreateProfile(profile).then(updatedProfile => {
-                    if (updatedProfile) {
-                        return self.validateAndSaveProfile(updatedProfile);
+                // Check whether the error was for firewall rule or not
+                if (self.connectionManager.failedUriToFirewallIpMap.has(uri)) {
+                    // Firewall rule error
+                    const clientIp = self.connectionManager.failedUriToFirewallIpMap.get(uri);
+
+                    // Check whether the azure account extension is installed and active
+                    if (self._vscodeWrapper.azureAccountExtensionActive) {
+                        // Sign in to azure account
+                        return self.promptForAccountSignIn().then(async (signedIn) => {
```

Comment:
Instead of having this huge amount of nested thens it would be greatly beneficial to just convert this into an async function for readability purposes...